File: 3960-frequency-balance-subarray/3960-frequency-balance-subarray.go
func getLength(nums []int) int {
	n, ans := len(nums), 0

	mp := make(map[int]int)
	idx := 0

	for i := 0; i < n; i++ {
		if _, ok := mp[nums[i]]; !ok {
			mp[nums[i]] = idx
			idx++
		}
		nums[i] = mp[nums[i]]
	}

	for i := 0; i+ans < n; i++ {
		length := n - i + 1
		if t := n - idx + 2; t < length {
			length = t
		}

		freq := make([]int, idx)
		freqFreq := make([]int, length)
		numCount, freqCount := 0, 0

		for j := i; j < n; j++ {
			val := freq[nums[j]] + 1
			freq[nums[j]] = val

			if val == 1 {
				numCount++
			}

			if freqFreq[val] == 0 {
				freqCount++
			}
			freqFreq[val]++

			if val != 1 {
				freqFreq[val-1]--
				if freqFreq[val-1] == 0 {
					freqCount--
				}
			}

			if ans > j-i+1 {
				continue
			}

			if numCount == 1 ||
				(freqCount == 2 &&
					(((val&1) == 0 && freqFreq[val>>1] != 0) ||
						((val<<1) < length && freqFreq[val<<1] != 0))) {
				ans = j - i + 1
			}
		}
	}

	return ans
}

File: 3960-frequency-balance-subarray/3960-frequency-balance-subarray.c
int getLength(int* nums, int numsSize) {
    int n = numsSize, ans = 0;

    int *vals = (int*)malloc(n * sizeof(int));
    int idx = 0;

    for (int i = 0; i < n; i++) {
        int j;
        for (j = 0; j < idx; j++)
            if (vals[j] == nums[i]) break;

        if (j == idx)
            vals[idx++] = nums[i];

        nums[i] = j;
    }

    for (int i = 0; i + ans < n; i++) {
        int len = (n - i + 1 < n - idx + 2) ? n - i + 1 : n - idx + 2;
        int *freq = (int*)calloc(idx, sizeof(int));
        int *freqFreq = (int*)calloc(len, sizeof(int));

        int numCount = 0, freqCount = 0;

        for (int j = i; j < n; j++) {
            int val = ++freq[nums[j]];

            if (val == 1) numCount++;

            if (freqFreq[val]++ == 0) freqCount++;
            if (val != 1 && --freqFreq[val - 1] == 0) freqCount--;

            if (ans > j - i + 1) continue;

            if (
                numCount == 1 ||
                (
                    freqCount == 2 &&
                    (
                        ((val & 1) == 0 && freqFreq[val >> 1] != 0) ||
                        ((val << 1) < len && freqFreq[val << 1] != 0)
                    )
                )
            ) {
                ans = j - i + 1;
            }
        }

        free(freq);
        free(freqFreq);
    }

    free(vals);
    return ans;
}